rollingpin/elasticsearch.py

```diff
@@ -62,56 +62,131 @@ def index_doc(self, doc):
             )
 
+    def update_doc(self, id_, updated_fields):
+        """ Update a document in Elasticsearch
+        :param updated_fields: dictionary values to update
+        :return Deferred
+        """
+        # The twisted web client chokes on unicode URLs, so we'll coerce this
+        # back to a string
+        update_endpoint = str("%s/%s/_update" % (self.endpoint, id_))
+
+        # Unlike indexing, for updates we have to put our updated fields inside
+        # a "doc" key.
+        #
+        # https://www.elastic.co/guide/en/elasticsearch/guide/current/partial-updates.html
+        doc = {'doc': updated_fields}
+
+        with swallow_exceptions("elasticsearch", self.logger):
+            agent = Agent(reactor)
+            body = JSONBodyProducer(json.dumps(doc))
+            return agent.request(
+                'POST',
+                update_endpoint,
+                Headers({'User-Agent': ['rollingpin']}),
+                body,
+            )
+
+    def build_sync_doc(self, sync_info):
+        timestamp_in_milliseconds = int(time.time()) * 1000
+
+        # We aren't aware of the commit ID that is being deployed at initial
+        # deploy time.  Once the build sync has happened, we have this
+        # information, so we go back and update the original document with a
+        # list of the sync targets (i.e. commit IDs being deployed).
+        sync = ["%s@%s" % (k, v.get('token', '')[:7])
+                for k, v in sync_info.iteritems()]
+        return {
+            'timestamps': {
+                'build': {
+                    'sync': timestamp_in_milliseconds,
+                },
+            },
+            'sync_targets': ', '.join(sync),
+        }
+
     def deploy_start_doc(self):
         timestamp_in_milliseconds = int(time.time()) * 1000
         return {
             'id': self.deploy_name,
+
+            # Once we change the mapping and reindex the data we can remove
+            # this redundant timestamp field.
             'timestamp': timestamp_in_milliseconds,
+
+            'timestamps': {
+                'deploy': {
+                    'start': timestamp_in_milliseconds,
+                },
+            },
             'components': self.components,
             'deployer': getpass.getuser(),
             'command': self.command_line,
             'hosts': self.hosts,
             'host_count': len(self.hosts),
-            'event_type': 'deploy.begin',
         }
 
     def deploy_abort_doc(self, reason):
         timestamp_in_milliseconds = int(time.time()) * 1000
         return {
-            'id': self.deploy_name,
-            'timestamp': timestamp_in_milliseconds,
-            'reason': reason,
-            'components': self.components,
-            'event_type': 'deploy.abort',
+            'timestamps': {
+                'deploy': {
+                    'abort': timestamp_in_milliseconds,
+                },
+            },
+            'abort_reason': reason,
```

Comment:
Yeah, this structure will make us unable to display that information at all though, since we don't have a way to iterate through arrays and decide what information to display.  So if we wanted to start graphing aborts or ends or some other arbitrary event, we're pretty much just stuck with deploy starts.